### DISCUSSION TITLE
Upgrading Dockersized Netbox to 3.3.8

### DISCUSSION THREAD

**Discussion by orenr2301** [netbox-community/netbox#11043] [Q&A]
Hello Everyone, 

I have recently installed netbox 3.3.7 via docker compose

The 3.3.8  is out but there is no manual how to upgrade composed netbox containers 
I  have went through the manual gui  from   https://docs.netbox.dev/en/stable/installation/upgrading/

The container to be upgraded through is the netbox-docker-netbox-worker if trying to do it manually

some files from tar.gz is not there by default - venv is  even not there 

Can you put a handover for how to do it? 

Also you didnt merge this version yet so when asked to clone the git repo it says its already says " Branch already updated" 

 **Because then it will be easy just to do the followings steps:** 
2. take sql dump : 
`docker exec -t POSTGRES_CONTAINER_ID pg_dumpall -c -U netbox > dump_`date +%d-%m-%Y"_"%H_%M_%S`.sql`
3. Go to /opt/netbox-docker/ {go to the netbox docker git directory} then Stop netbox-docker
`docker-compose down`

4. Rename docker-composer.yml and configuration/ldap_config.py 

6. Pull the latest netbox from the git repository
7. `git pull origin release`
`docker-compose rm -fs netbox netbox-worker postgres`
`docker-compose pull`

8.Clean up old data
`docker volume rm netbox-docker_netbox-postgres-data`
`docker volume rm netbox-docker_netbox-static-files`

9. Restore postgres database content
`docker-compose up postgres'
'cat your_dump.sql | docker exec -i POSTGRES_CONTAINER_ID psql -U netbox`

10. Finally, bring everything up a
`docker-compose up -d`

**Comment by candlerb** [comment]
This is really a more general question, "how do I upgrade applications deployed using docker"?

The short answer is, you don't.  You *never* modify or upgrade an application within an existing container.  Instead, you delete and recreate the container from a new image, which contains the new version of the application.  The application needs to be deployed in such a way that its persistent state is kept in a data volume, so when the new version starts up, it will find the old data.

If you look at the `docker-compose.yml` file, you'll see that it starts multiple containers.  One of those is postgres:

```
  postgres:
    image: postgres:14-alpine
    env_file: env/postgres.env
    volumes:
    - netbox-postgres-data:/var/lib/postgresql/data
```

As it happens, you don't want to upgrade Postgres at the moment, only Netbox.  But if you did, you can see that the data directory `/var/lib/postgresql/data` is stored in the named volume `netbox-postgres-data`, so the newly-deployed Postgres container would pick up the existing data.

Now, 90% of Netbox's state is kept in Postgres, so having Postgres data preserved solves most of the problem.  The rest is in Netbox's media directory (for uploaded images), any scripts and reports you create, and I suppose also redis if you're worried about preserving history of job executions.

Again, looking in `docker-compose.yml`, you can see that the Netbox container attaches multiple data volumes for those directories which need to be preserved:

```
services:
  netbox: &netbox
    image: netboxcommunity/netbox:${VERSION-v3.2-2.0.0}
...
    volumes:
    - ./startup_scripts:/opt/netbox/startup_scripts:z,ro
    - ./initializers:/opt/netbox/initializers:z,ro
    - ./configuration:/etc/netbox/config:z,ro
    - ./reports:/etc/netbox/reports:z,ro
    - ./scripts:/etc/netbox/scripts:z,ro
    - netbox-media-files:/opt/netbox/netbox/media:z
```

So in summary: you just redeploy the brand new version of Netbox, using a new container image which was built from Netbox v3.3.8, and Bob's your uncle.

**Reply by cyberndj** [reply]
Howdy,
https://github.com/netbox-community/netbox-docker/wiki/Updating has a resource to help you perform an upgrade for your containers.

Same WiKi page also has a guide to making a custom image if you want to have any plugins be persistent. On my environment, I set my custom docker compose to set the specific NetBox version to pull/create my persistent image w/plugins, then recreate my containers to get the update.

**Comment by orenr2301** [comment]
Ok , some how i understood that i need to downloads gzip file and do some manually stuff or reclone the git repo.

But of course that i need to take the new netbox-image in the compose. like daaa. 

Thanks guys for take me out of the stupidity zone :)
